### PR TITLE
Timestamps for chat

### PR DESCRIPTION
**Pull request by cjb**
@keybase/react-hackers 

Here's a PR for chat timestamps!  Changes:

* On adding a new message ourselves, or receiving a new message, or going through a list of old messages:

* We go through them pairwise, adding a timestamp if the current one is more than a constant (currently 15 mins) newer than the previous one.

* A timestamp is a new kind of Message.  It's the only type without a MessageID (client generated rather than server generated).  I wasn't sure whether that was a good idea, it seems better than the alternatives (like annotating each Message that should also display a timestamp).

* For formatting the timestamps, the messages view uses a different moment formatting string than the timestamps in the conversation list view.  There's less screen space for the conversation list view ones, so it makes sense to provide more detail here.

### CODE REVIEW

File: shared/actions/chat.js [hunk 1]
```diff
@@ -172,11 +190,29 @@ function * _postMessage (action: PostMessage): SagaGenerator<any, any> {
       message: new HiddenString(action.payload.text.stringValue()),
       followState: 'You',
     }
+
+    // Should we add a timestamp before our new message?
+    // TODO short-term if we haven't seen this in the conversation list we'll refresh the inbox. Instead do an integration w/ gregor
+    const conversationStateSelector = (state: TypedState) => state.chat.get('conversationStates', Map()).get(conversationIDKey)
```
Comment (chrisnojima) [inline shared/actions/chat.js]: how do you get into this state? you're making messages in the cli but not on the chat screen?

File: shared/actions/chat.js [hunk 2]
```diff
@@ -172,11 +190,29 @@ function * _postMessage (action: PostMessage): SagaGenerator<any, any> {
       message: new HiddenString(action.payload.text.stringValue()),
       followState: 'You',
     }
+
+    // Should we add a timestamp before our new message?
```
Comment (chrisnojima) [inline shared/actions/chat.js]: We should add a timestamp on new messages

File: shared/actions/chat.js [hunk 3]
```diff
@@ -332,6 +392,19 @@ function _threadToPagination (thread) {
   }
 }
 
+function _maybeAddTimestamp (message: Message, prevMessage: Message): MaybeTimestamp {
+  if (message.type !== 'Text' || prevMessage.type !== 'Text') {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: We should likely always do the timestamp, not just between text messages

File: shared/chat/conversation/list.desktop.js [hunk 4]
```diff
@@ -54,7 +54,7 @@ class ConversationList extends Component<void, Props, State> {
       // We want a stable key -- messages have an outboxID but no messageID,
       // then later gain a messageID.  So if we prefer outboxIDs to messageIDs
       // for the key, every row keeps its key.
-      const id = message && (message.outboxID || message.messageID)
+      const id = message && (message.outboxID || message.messageID || message.timestamp)
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: Should we explicitly check if type==='timestamp' and use timestamp vs this || fallback?

File: shared/chat/conversations-list/index.desktop.js [hunk 5]
```diff
@@ -1,27 +1,12 @@
...
 import React from 'react'
-import moment from 'moment'
 import {Box, Text, Avatar, Icon, Usernames} from '../../common-adapters'
 import {globalStyles, globalColors} from '../../styles'
 import {participantFilter} from '../../constants/chat'
+import {formatTimeForConversationList} from '../../util/timestamp'
 
 import type {Props} from './'
 
-function _timestamp (time: number, nowOverride?: number): string {
```
Comment (chrisnojima) [inline shared/chat/conversations-list/index.desktop.js]: just fyi @gabriel  also moved this and renamed it something different

File: shared/actions/chat.js (hunk 1 above)
Comment (cjb) [inline shared/actions/chat.js]: Ah, I just brought this guard over, I've never hit that state, I'll remove it.

File: shared/actions/chat.js (hunk 3 above)
Comment (cjb) [inline shared/actions/chat.js]: > We should likely always do the timestamp, not just between text messages

In this case:

prevMessage = Timestamp
Message = Text

We wouldn't want to add another timestamp to make Timestamp => Timestamp => Text.  I'll change the test to be `if (message.type === 'Timestamp' || prevMessage.type === 'Timestamp') { return null }` here.

File: shared/chat/conversation/list.desktop.js (hunk 4 above)
Comment (cjb) [inline shared/chat/conversation/list.desktop.js]: Good idea, thanks, done.

**Comment by cjb** [comment]
@chrisnojima bump (ready for another look), thanks!

File: shared/actions/chat.js [hunk 6]
```diff
@@ -312,13 +360,24 @@ function * _loadMoreMessages (): SagaGenerator<any, any> {
 
   const yourName = yield select(usernameSelector)
   const messages = (thread && thread.thread && thread.thread.messages || []).map((message, idx) => _unboxedToMessage(message, idx, yourName, conversationIDKey)).reverse()
+  let newMessages = []
+  messages.forEach((message, idx) => {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: Don't need to make this change but you can use reduce for cases like this instead of mutating newMessages

File: shared/chat/conversation/list.desktop.js [hunk 7]
```diff
@@ -248,7 +255,7 @@ class ConversationList extends Component<void, Props, State> {
 }
 
 class CellSizeCache extends defaultCellMeasurerCellSizeCache {
-  _indexToID: (index: number) => number;
+  _indexToID: (index: number) => ?number;
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: Why this change?

Comment (cjb) [inline shared/chat/conversation/list.desktop.js]: @chrisnojima 
Without this change:
```
chat/conversation/list.desktop.js:45
 45:     this._cellCache = new CellSizeCache(this._indexToID)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructor call
 71:       return id
                  ^^ possibly uninitialized variable. This type is incompatible with
258:   _indexToID: (index: number) => number;
                                      ^^^^^^ number
```
The code is:
```js
      let id
      if (message) {
        if (message.type === 'Timestamp') {
          id = message.timestamp
        } else {
          // We want a stable key -- messages have an outboxID but no
          // messageID, then later gain a messageID.  So if we prefer
          // outboxIDs to messageIDs for the key, every row keeps its key.
          id = message && (message.outboxID || message.messageID)
        }
      }
      if (id == null) {
        console.warn('id is null for index:', messageIndex)
      }
      return id
    }
```
So the function's capable of returning null (even though we never want it to), but the flow type was requiring a number.

**Comment by chrisnojima** [comment]
But wasn't this logically the same before the time stamps

**Comment by cjb** [comment]
> But wasn't this logically the same before the time stamps

Yeah.  The old code was:
```js
  const id = message && (message.outboxID || message.messageID)
```
so perhaps flow felt capable of proving that message would always be a Message type (and so id would always be non-null) before, and now it doesn't feel that way for some reason.  I think the new type is more correct than the old type!

**Comment by chrisnojima** [comment]
Hmmmm. Ok I'm not so worried about it at this point. Ship it!
